File: leetcode_511.cpp
class Solution {
public:
  vector<int> preorderTraversal(TreeNode *root) {
    vector<int> result;
    const TreeNode *p;
    stack<const TreeNode *> s;

    p = root;
    if (root != nullptr) s.push(p);

    while (!s.empty()) {
      p = s.top();
      s.pop();
      result.push_back(p);
      if (p->right != nullptr) s.push(p->right);
      if (p->left != nullptr) s.push(p->left);
    }

    return result;
  }
}

File: leetcode_511.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
  public ArrayList<Integer> preorderTraversal(TreeNode root) {
    Stack<TreeNode> stack = new Stack<TreeNode>();
    ArrayList<Integer> result = new ArrayList<Integer>();
    if (root != null) {
      stack.push(root);
    }
    while (stack.size() != 0) {
      TreeNode node = stack.pop();
      result.add(node.val);
      if (node.right != null) {
        stack.push(node.right);
      }
      if (node.left != null) {
        stack.push(node.left);
      }
    }

    return result;        
  }
}
